File: web/di_core/domain/Model/Column/Implement/StringColumn.ts
/*
 * @author: tvc12 - Thien Vi
 * @created: 5/29/21, 4:33 PM
 */

import { ColumnType } from '@core/domain/Model/Column/Implement/ColumnType';
import { Column } from '@core/domain/Model/Column/Column';
import { Expression } from '@core/domain/Model/Column/Expression/Expression';

export class StringColumn extends Column {
  className = ColumnType.string;
  name!: string;
  displayName!: string;
  description?: string;
  defaultValue?: string;
  isNullable: boolean;
  isEncrypted: boolean;
  isPrivate: boolean;
  defaultExpression?: Expression;

  constructor(
    name: string,
    displayName: string,
    isNullable: boolean,
    isEncrypted: boolean,
    isPrivate: boolean,
    description?: string,
    defaultValue?: string,
    defaultExpression?: Expression
  ) {
    super();
    this.name = name;
    this.displayName = displayName;
    this.description = description;
    this.defaultValue = defaultValue;
    this.isNullable = isNullable || false;
    this.isEncrypted = isEncrypted || false;
    this.isPrivate = isPrivate || false;
    this.defaultExpression = defaultExpression;
  }

  static fromObject(obj: StringColumn): StringColumn {
    const defaultExpression = obj.defaultExpression ? Expression.fromObject(obj.defaultExpression) : void 0;

    return new StringColumn(obj.name, obj.displayName, obj.isNullable, obj.isEncrypted, obj.isPrivate, obj.description, obj.defaultValue, defaultExpression);
  }

  static empty(): StringColumn {
    return new StringColumn('', '', true, false, false, '', '');
  }
}

File: server/src/main/scala/co/datainsider/schema/domain/column/StringColumn.scala
package co.datainsider.schema.domain.column

@SerialVersionUID(20200715L)
case class StringColumn(
    name: String,
    displayName: String,
    description: Option[String] = None,
    defaultValue: Option[String] = None,
    isNullable: Boolean = false,
    isEncrypted: Boolean = false,
    defaultExpr: Option[String] = None,
    defaultExpression: Option[DefaultExpression] = None
) extends Column {
  override def copyTo(name: String, displayName: String): Column = this.copy(name = name, displayName = displayName)

  override def copyTo(
      name: String,
      displayName: String,
      description: Option[String],
      isNullable: Boolean,
      defaultExpression: Option[DefaultExpression],
      isEncrypted: Boolean
  ): Column = {
    this.copy(
      name = name,
      displayName = displayName,
      description = description,
      isNullable = isNullable,
      defaultExpression = defaultExpression,
      isEncrypted = isEncrypted
    )
  }
}
